### PR TITLE
Accept all types of revision specifiers

### PR DESCRIPTION
**Pull request by ca-johnson**
Docs on revision specifiers:
https://www.perforce.com/manuals/v17.1/cmdref/Content/CmdRef/filespecs.synopsis.using_revision_specifiers.html?TocPath=File%20Specifications%7C_____3

* move logic out of checkout.py and into buildkite.py (since its buildkite-context-specific)
* handle more revision specifiers than a plain changelist number
* resolves internal issue ENG-2799

testing strategy:
integration test

### CODE REVIEW

File: python/checkout.py [hunk 1]
```diff
@@ -19,15 +18,10 @@ def main():
     repo = P4Repo(**config)
 
     revision = get_build_revision()
-    if not revision.isdigit():
-        # Resolve 'HEAD' or ignore git sha and find a concrete revision
+    if revision is None:
```
Comment (ca-johnson) [inline python/checkout.py:21]: Now we only resolve to head if we completely fail to get a revision from buildkite context

File: python/buildkite.py [hunk 2]
```diff
@@ -94,8 +95,22 @@ def set_build_changelist(changelist):
...
+    """Get a p4 revision for the build from buildkite context"""
+    revision = get_metadata(__REVISION_METADATA__)
+    if revision:
+        return revision
+
+    revision = os.environ['BUILDKITE_COMMIT'] # HEAD, user-defined revision or git-sha
+    # Convert bare changelist number to revision specifier
+    # Note: Theoretically, its possible for all 40 characters of a git sha to match this.
+    #       In practice, the inconvenience of forcing users to always include '@' outweighs this risk (~1 in 7 billion)
+    if re.match(r'^\d*$', revision):
```
Comment (stefan-improbable) [inline python/buildkite.py]: I like the `revision.isdigit()` personally more, not sure if we can use it here. Up to you.

**Review by stefan-improbable — APPROVED** [review]
lgtm % nit?

File: python/buildkite.py (hunk 2 above)
Comment (ca-johnson) [reply python/buildkite.py]: agreed, good catch thanks

**Pull request merged**